RONDB-832: More work on integrating Rondis into RDRS2

mronstro opened the pull request.

mikaelronstrom committed:
RONDB-832: Integration of Rondis into RDRS2

Made it possible to run Rondis as a standalone process or as library
in RDRS2 that opens its own REDIS port to users.

Implemented callback functions to get and return Ndb objects from the
pool common with RDRS2.

Implemented a callback function to exit the process in failure situation.

svenssonaxel committed:
RONDB-832 Review comments for #641

svenssonaxel committed:
Add rdrs2 config variable .RondisConfig.Databases[].DirtyINCRDECR

mikaelronstrom committed:
RONDB-832: Further improvements of Rondis

Added support for dirty_flag in configuration
Added possibility to switch between databases.

svenssonaxel commented on storage/ndb/rest-server2/server/src/main.cc:
```diff
@@ -293,17 +322,28 @@ int main(int argc, char *argv[]) {
-    g_rondis_thread = NewDispatchThread(globalConfigs.rondis.serverPort,
+    g_database_index = (int*)malloc(sizeof(int) * num_rondis_threads);
+    memset(g_database_index, 0, sizeof(int) * num_rondis_threads);
+    bool dirty_incr_decr_flag[MAX_NUM_DATABASES];
```
`MAX_NUM_DATABASES` should be used in `config_structs_def.hpp` in this line:
`PROBLEM(numDatabases > 1000, "Number of rondis databases too high")`

mronstro replied:
ok will do

mronstro replied:
Done

svenssonaxel commented on storage/ndb/rest-server2/server/src/main.cc:
```diff
@@ -293,17 +322,28 @@ int main(int argc, char *argv[]) {
+    for (Uint32 i = 0; i < globalConfigs.rondis.numDatabases; i++) {
+      dirty_incr_decr_flag[i] = globalConfigs.rondis.databases[i].dirtyIncrDecr;
+    }
+    g_rondis_thread = NewDispatchThread(globalConfigs.rondis.serverIP,
```
You might want to remove:
```
 PROBLEM(serverIP != "0.0.0.0",
         "Setting rondis server IP to anything else than 0.0.0.0 is not yet"
         " implemented.")
```

mronstro replied:
ok will do

mronstro replied:
Done

svenssonaxel commented on storage/ndb/rest-server2/server/src/main.cc:
```diff
@@ -293,17 +322,28 @@ int main(int argc, char *argv[]) {
+    setup_ndb_connection_for_rondis(get_ndb_object,
+                                    return_ndb_object,
+                                    exit_on_rondis_error,
+                                    num_rdrs_threads,
```
Are you sure? Or should it be `num_rdrs_threads + num_purge_threads`?

mronstro replied:
I am sure, first comes the RDRS threads, second the Rondis threads and last the purge threads

svenssonaxel commented on storage/ndb/src/rondis/src/rondb.cc:
```diff
@@ -150,7 +242,7 @@ int rondb_redis_handler(const pink::RedisCmdArgsType &argv,
     }
     response->append("+PONG\r\n");
-  } else if (argv[0] == "ECHO") {
+  } else if (strcasecmp(command, "echo") == 0) {
```
Perhaps you could optimize this by turning command into lowercase once and then do `strcmp`.

mronstro replied:
Later, need a much more clever way of selecting command using hash, so no need to micro optimise before that

mikaelronstrom committed:
RONDB-832: A lot of fixes to integrate Rondis into RonDB and into RDRS server, even some statistics update of NDB API

mikaelronstrom committed:
RONDB-832: Ensure that INCR/DECR doesn't make the row disappear immediately

mikaelronstrom committed:
RONDB-832: Added include for assert

mikaelronstrom committed:
RONDB-832: Fixed TTL impact on SET, MSET, HSET, HMSET command

mikaelronstrom committed:
RONDB-832: Introduced support for 16 databases in Rondis

mikaelronstrom committed:
RONDB-832: Support select database in Rondis

mikaelronstrom committed:
RONDB-832: Added prometheus support to Rondis

mikaelronstrom committed:
RONDB-832: More work on configuration handling, preparing for better handling of expiration of rows

mikaelronstrom committed:
RONDB-832: Preparations to support more complex SET operations

mikaelronstrom committed:
RONDB-832: First step to proper handling of TTL and SET NX and SET XX

mikaelronstrom committed:
RONDB-832: Handle Keep TTL feature in Rondis

mikaelronstrom committed:
RONDB-832: Implemented SET key val EXAT unix_time and PXAT (milliseconds)

mikaelronstrom committed:
RONDB-832: First step towards support SET -GET for Rondis

mikaelronstrom committed:
RONDB-832: Introduce NULLable expiry_date and fix a number of minor issues

mikaelronstrom committed:
RONDB-832: Fixed SET key val EXAT/PXAT feature

mikaelronstrom committed:
RONDB-832: Got SET .. GET working

mikaelronstrom committed:
RONDB-832: Added support for STRLEN command

mikaelronstrom committed:
RONDB-832: Support GETRANGE command in Rondis

mikaelronstrom committed:
RONDB-832: Handled SETRANGE with changes in key row only

mikaelronstrom committed:
RONDB-832: Handling key row for SETRANGE where range reachees beyond key row

mikaelronstrom committed:
RONDB-832: Handle value rows for SETRANGE

mikaelronstrom committed:
RONDB-832: Fix assert include

mikaelronstrom committed:
RONDB-861: Added interpreter instructions

- Added input parameters to interpreted programs
- Prepare for binary search instructions
- First step introducing sort algorithms in interpreter
- More work on sort algorithms in interpreter

mikaelronstrom committed:
RONDB-861: More work on new interpreter instructions

mikaelronstrom committed:
RONDB-832: Simplified handling of SETRANGE for value rows

mikaelronstrom committed:
RONDB-832: Fixes a minor bug, added printout of start info on Rondis and added a script to create Rondis tables

mikaelronstrom committed:
RONDB-832: First stab at more complex SETRANGE calls

mikaelronstrom committed:
RONDB-832: Many fixes to SETRANGE command

mikaelronstrom committed:
RONDB-832: Disabled some debugging and added a bit more documentation of interpreter and fixed some error checks
